File: solutions/javascript/ReverseNodesink-Group/ReverseNodesink-Group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    var copy = head;
    if (k <= 1 || !head) {
        return head;
    }
    var a = [];
    var b = [];
    while (copy != null) {
        a.push(copy.val);
        copy = copy.next;
    }
    if (a.length < k) {
        return head;
    }
    for (var i = 0; i < a.length; i+=k) {
        if (a.slice(i, i+k).length == k) {
            b = b.concat(a.slice(i, i+k).reverse());
        } else {
            b = b.concat(a.slice(i, i+k));
        }
    }
    var res = new ListNode(b[0]);
    copy = res;
    for (var j = 1; j < b.length; j++) {
        copy.next = new ListNode(b[j]);
        copy = copy.next;
    }
    return res;
};

File: solutions/java/ReverseNodesink-Group/ReverseNodesink-Group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        //列表长度为0或1时直接返回
        if (null == head || null == head.next) {
            return head;
        }
        ListNode left = head;
        ListNode right = head;
        int counter = k;

        while (counter > 1) {
            right = right.next;
            counter--;
            if (null == right) {
                return head;
            }
        }

        head = right;
        ListNode last = left;
        ListNode next = right.next;

        while (null != right) {
            while (left != right) {
                ListNode tmp = left;
                left = left.next;
                tmp.next = right.next;
                right.next = tmp;
            }
            counter = k;
            left = next;
            right = next;
            while (counter > 1 && null != right) {
                right = right.next;
                counter--;
            }
            if (null != right) {
                next = right.next;
                last.next = right;
                last = left;
            }
        }
        return head;
    }
}